### ISSUE TITLE
Warning: token pasting of ',' and __VA_ARGS__ is a GNU extension

### ISSUE THREAD

**Issue by chrhaase** [sudara/melatonin_perfetto#27]
Hi, building with `juce_recommended_warning_flags` I get this warning:

```
_deps/melatonin_perfetto-src/melatonin_perfetto/melatonin_perfetto.h:249:69: warning: token pasting of ',' and __VA_ARGS__ is a GNU extension [-Wgnu-zero-variadic-macro-arguments]
            TRACE_EVENT ("dsp", perfetto::StaticString (pf.data()), ##__VA_ARGS__)
```

I had the same issue with my own logging macros and simply removing the `##` solved it. I'm building with Clang though, so I thought those might would be needed to make gcc happy, but [maybe not actually](https://godbolt.org/z/s96YMnTdv)...

**Comment by davidspry** [comment]
I believe this is a portable replacement, although it [requires C++20](https://en.cppreference.com/w/cpp/preprocessor/replace):
```cpp
TRACE_EVENT ("dsp", perfetto::StaticString (pf.data()) __VA_OPT__(,) __VA_ARGS__)
```

**Comment by chrhaase** [comment]
Yes! That's what I've ended up doing as well with my logging macros.
On MSVC you need to compile with `/Zc:preprocessor`.

**Comment by sudara** [comment]
Thanks for this! 

Do either of these fix the warning on C++17? 

I'm on C++20 for my own project, but this library has to stay compatible with C++17 (JUCE 8's and Perfetto's current minimum). Honestly, the preprocessor stuff has been quite a headache, so would personally selfishly love it to be > C++20...

I wonder if the only answer to staying on C++17 would be disabling the warning at the file/module level..

**Comment by chrhaase** [comment]
Yeah, I'm also on C++20 here, but probably sensible to keep C++17 compatibility.
I don't think my first "fix" works at all actually :)

I've been playing with this on [godbolt](https://godbolt.org/z/boxMnsEoo) and I guess a solution could be using `__VA_OPT__` for C++20 and then just suppressing the warning on Clang/C++17.
I haven't checked older compiler versions, I don't know if there could be some dragons there..

**Comment by sudara** [comment]
> I've been playing with this on [godbolt](https://godbolt.org/z/sMKbWzaxM) and I guess a solution could be using __VA_OPT__ for C++20 and then just suppressing the warning on Clang/C++17.

Nice! Yeah, that seems sensible to me...
